cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AbstractDistributedProgramRunner.java

```diff
@@ -366,6 +368,11 @@ private TwillPreparer addSecureStore(TwillPreparer preparer, LocationFactory loc
       HBaseTokenUtils.obtainToken(hConf, credentials);
     }
 
+    if (cConf.getBoolean(Constants.Explore.EXPLORE_ENABLED) && UserGroupInformation.isSecurityEnabled()) {
+      HiveTokenUtils.obtainToken(credentials);
```

Comment:
Is the master services and all program runners has access to the same set of delegation tokens? If that's the case, we should refactor the program runner a bit to call `TokenSecureStoreUpdater` to get a set of delegation tokens just as how it is done in master service.